### PR TITLE
Testing: Add support for running tests after bootstrapping the dev environment; rucio#7928

### LINKED ISSUE #7928: Testing: Introduce a helper functionality for local testing
### Description

It would be very helpful (for dev/debugging efficiency) to be able to run easily the tests we want, locally.

In general the following features would be more than welcome:
- Be able to repeat the testing quickly without having to reconfigure the environment (e.g. due to Alembic problems).
- Be able to execute also some autotest module.
- Do that automatically with low effort (e.g. avoid having to construct complex commands).
- Clear documentation.

I am attaching this as a sub-issue to #7667 since it is quite closely related to that.

### Steps to reproduce

.

### Rucio Version

Server: 38.0.0rc2

### Additional Information

_No response_

### PR DESCRIPTION
**Pull request by Geogouz**
**Draft Notice:** This works but kept frozen until we refactor our CI. Then I will re-adjust.
---

Fixes #7928

**Summary**
This PR extends the `./tools/bootstrap_dev.sh` script to optionally run (using -t/--test) a selected test suite immediately after setup. The script lists the available test options, validates the chosen test number, and starts/stops any required services for the selected scenario automatically when autotests are used. Ideally the number of available tests should match at the end the one we have in our CI (e.g. 31). Keep in mind that this PR depends on #7910 so we should merge that first (and then I could rebase).

**Usage**
- Run a specific test right after bootstrapping (e.g. `./tools/bootstrap_dev.sh -t 1` - for using the local test suite)
- Tests run against a clean database; the dev DB volume is reset to avoid state carryover.
- The available/offered autotests are detected automatically based on our matrix.
- The -f/--filter option can be used with -t/--test to pass any pytest-style filter such as 'tests/file.py::TestClass::test_method' to run a subset of tests.
    - With autotests (i.e. when `-t > 1` is selected), this should work fine. However, when `-t 1` is used (i.e. local test suite), since currently the default docker-compose profile is only allowed to be spinned up for that, you should select a pytest-filter that is supposed to run with those default containers. This should ideally be changed in the future so that the proper profile is automatically selected. But this needs a bit of discussion of how we should do it.
- The -c/--cache-use option skips rebuilding images if a cached version exists.
- In the future, I can also make an official instructions guide about this script's usage (to be placed on our documentation page), but for now, the complete instructions can be found in more detail at the top of `./tools/bootstrap_dev.sh` itself.

    _Example:_ 
    Run a single pytest within `multi_vo (dist alma9, py 3.9, db postgres14)` test suite using a filter:
    `./tools/bootstrap_dev.sh --test 9 --filter tests/test_scope.py::test_scope_duplicate`
    or just
    `./tools/bootstrap_dev.sh -t 9 -f tests/test_scope.py::test_scope_duplicate`

**Note**
This is a quite safe-to-merge PR since it affects only the `bootstrap_dev.sh` functionality which is used only by devs. However, since this convenience tool is still under active development and may occasionally break. If you encounter unexpected behavior, please open an issue so we can investigate and fix it. I already know couple edge-cases where upgrades are needed.

**Currently supported tests**
``` text
Available tests:
   1) run local test suite (tools/run_tests.sh)
   2) autotest: syntax (dist alma9, py 3.9)
   3) autotest: client_syntax (dist alma9, py 3.9)
   4) autotest: client (dist alma9, py 3.9, db postgres14)
   5) autotest: remote_dbs (dist alma9, py 3.9, db oracle, services influxdb_elastic)
   6) autotest: remote_dbs (dist alma9, py 3.9, db mysql8, services influxdb_elastic)
   7) autotest: remote_dbs (dist alma9, py 3.9, db postgres14, services influxdb_elastic)
   8) autotest: sqlite (dist alma9, py 3.9, db sqlite)
   9) autotest: multi_vo (dist alma9, py 3.9, db postgres14)
  10) autotest: syntax (dist alma9, py 3.10)
  11) autotest: client_syntax (dist alma9, py 3.10)
  12) autotest: client (dist alma9, py 3.10, db postgres14)
  13) autotest: remote_dbs (dist alma9, py 3.10, db oracle, services influxdb_elastic)
  14) autotest: remote_dbs (dist alma9, py 3.10, db mysql8, services influxdb_elastic)
  15) autotest: remote_dbs (dist alma9, py 3.10, db postgres14, services influxdb_elastic)
  16) autotest: sqlite (dist alma9, py 3.10, db sqlite)
  17) autotest: multi_vo (dist alma9, py 3.10, db postgres14)
  18) integration workflow: integration-test (dist alma9, py 3.9)

Run with './tools/bootstrap_dev.sh --test <number>' or './tools/bootstrap_dev.sh -t <number>' to execute one automatically. Do not combine with -p/--profile; tests start required services themselves.
```

### CODE REVIEW

**Review by rdimaio — COMMENTED** [review]
I think that we should also make sure we keep this script maintainable and without too much complexity; it might make sense to keep the boostrap script solely for setting up everything, and a separate script for testing (which i guess is what the `run_tests.sh` script does), and having some logic where the run_tests script can be passed a flag to reset the overall environment if needed (maybe by calling the bootstrap_dev script)

I also think the number-based approach could be a bit confusing (i.e. the basic test suite is --test 1). Maybe being able to set this in words would be better (e.g. `--test basic`, `--test multi_vo)..)

**Comment by Geogouz** [comment]
Yes, I agree with making it easier to maintain because if when I get to include also the other tests and other functionalities, size might explode even more. I would need to see how this can be done.

Regarding the tests, my typical workflow (that is safe to possible future name changes etc) is to first get the list of possible tests using `./tools/bootstrap_dev.sh -h`, which reports:

```
...
Available tests:
   1) run local test suite (tools/run_tests.sh)
   2) autotest: syntax (dist alma9, py 3.9)
   3) autotest: client_syntax (dist alma9, py 3.9)
   4) autotest: client (dist alma9, py 3.9, db postgres14)
   5) autotest: remote_dbs (dist alma9, py 3.9, db oracle)
   6) autotest: remote_dbs (dist alma9, py 3.9, db mysql8)
   7) autotest: remote_dbs (dist alma9, py 3.9, db postgres14)
   8) autotest: sqlite (dist alma9, py 3.9, db sqlite)
   9) autotest: multi_vo (dist alma9, py 3.9, db postgres14)
  10) autotest: syntax (dist alma9, py 3.10)
  11) autotest: client_syntax (dist alma9, py 3.10)
  12) autotest: client (dist alma9, py 3.10, db postgres14)
  13) autotest: remote_dbs (dist alma9, py 3.10, db oracle)
  14) autotest: remote_dbs (dist alma9, py 3.10, db mysql8)
  15) autotest: remote_dbs (dist alma9, py 3.10, db postgres14)
  16) autotest: sqlite (dist alma9, py 3.10, db sqlite)
  17) autotest: multi_vo (dist alma9, py 3.10, db postgres14)
  18) integration workflow: integration-test (dist alma9, py 3.9)
```

and then just going with the test I am interested in like:

`./tools/bootstrap_dev.sh -t 5`

So I would like to keep the support for numbering but if you think, maybe I can add also support for call by some corresponding name which we will have to figure out a policy of how to write in an easy single word.

**Comment by rdimaio** [comment]
> Yes, I agree with making it easier to maintain because if when I get to include also the other tests and other functionalities, size might explode even more. I would need to see how this can be done.
> 
> Regarding the tests, my typical workflow (that is safe to possible future name changes etc) is to first get the list of possible tests using `./tools/bootstrap_dev.sh -h`, which reports:
> 
> ```
> ...
> Available tests:
>    1) run local test suite (tools/run_tests.sh)
>    2) autotest: syntax (dist alma9, py 3.9)
>    3) autotest: client_syntax (dist alma9, py 3.9)
>    4) autotest: client (dist alma9, py 3.9, db postgres14)
>    5) autotest: remote_dbs (dist alma9, py 3.9, db oracle)
>    6) autotest: remote_dbs (dist alma9, py 3.9, db mysql8)
>    7) autotest: remote_dbs (dist alma9, py 3.9, db postgres14)
>    8) autotest: sqlite (dist alma9, py 3.9, db sqlite)
>    9) autotest: multi_vo (dist alma9, py 3.9, db postgres14)
>   10) autotest: syntax (dist alma9, py 3.10)
>   11) autotest: client_syntax (dist alma9, py 3.10)
>   12) autotest: client (dist alma9, py 3.10, db postgres14)
>   13) autotest: remote_dbs (dist alma9, py 3.10, db oracle)
>   14) autotest: remote_dbs (dist alma9, py 3.10, db mysql8)
>   15) autotest: remote_dbs (dist alma9, py 3.10, db postgres14)
>   16) autotest: sqlite (dist alma9, py 3.10, db sqlite)
>   17) autotest: multi_vo (dist alma9, py 3.10, db postgres14)
>   18) integration workflow: integration-test (dist alma9, py 3.9)
> ```
> 
> and then just going with the test I am interested in like:
> 
> `./tools/bootstrap_dev.sh -t 5`
> 
> So I would like to keep the support for numbering but if you think, maybe I can add also support for call by some corresponding name which we will have to figure out a policy of how to write in an easy single word.

The number system looks good now that I see your overall workflow, I think it makes sense to keep it like this

File: etc/docker/dev/docker-compose.yml [hunk 1]
```diff
@@ -61,6 +61,13 @@ services:
       - REQUESTS_CA_BUNDLE=/etc/pki/tls/certs/ca-bundle.crt
       - RUCIO_SOURCE_DIR=/rucio_source
       - RDBMS
+    depends_on:
+      oracle:
+        condition: service_healthy
+        required: false
+      ruciodb:
+        condition: service_healthy
+        required: false
```
Comment (rdimaio) [inline etc/docker/dev/docker-compose.yml:71]: https://github.com/rucio/rucio/pull/7910 has been merged, please rebase

**Comment by Geogouz** [comment]
I am temporarily switching this to a draft because I need to double check how #7782 would affect it. So let's prioritise the PR from @maany first and I will adapt current PR as/if needed (maybe none is needed but let me double check first).

**Comment by maany** [comment]
Roger that 😎

**Comment by Geogouz** [comment]
I generated and attach below a draft documentation page guiding developers how to use the tool I want to introduce in this PR. To get an idea whether this can be useful to the community (this is what I use basically for local dev).

**Couple things to note:**
- The idea here is to centralise in a single robust script, all functionalities that someone might need for local development to use, offering those in an as easy way as possible. This includes all tests we currently run on our Github CI, plus debugging with breakpoints.
- Another strong point for using "bootstrap_dev.sh" is that it handles dependency upgrades when you bootstrap your `master` stack. This step solves the mismatch between the container image (built from the latest released tag) and the local checkout on master, which can depend on newer or different packages than the baked image provides. Currently, without this guard, bringing up services or running tests could fail with missing or outdated Python dependencies.
- I have already refactored the script into different modules but I will wait for merging the work of @maany first in order to adjust what needs to be adjusted in this tool.
- The way I have currently refactored the script will allow (as in second phase) to have various scripts involved in testing (and scattered around) to be integrated here and eventually removed (to reduce the amount of scripts we have).

@rdimaio @maany @bari12 @0xquark @mgajek-cern 

[bootstrap_dev.sh Usage Guide.pdf](https://github.com/user-attachments/files/22751770/bootstrap_dev.sh.Usage.Guide.pdf)

**Comment by dchristidis** [comment]
As part of an effort to keep the number of pull requests under control, your PR will be closed due to:

* The PR is in draft mode and there has been no activity for over seven days.

Please be advised that this is not a definitive rejection of your effort. Feel free to reopen it once the aforementioned reason has been addressed.
